pkg/credentialprovider/plugin/config.go

```diff
@@ -0,0 +1,123 @@
+	if configPath == "" {
+		return nil, fmt.Errorf("credential provider config path is empty")
+	}
+
+	data, err := ioutil.ReadFile(configPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read external registry credential provider configuration from %q: %v", configPath, err)
+	}
+
+	config, err := decode(data)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding config %s: %v", configPath, err)
+	}
+
+	return config, nil
+}
+
+// decode decodes data into the internal CredentialProviderConfig type.
+func decode(data []byte) (*kubeletconfig.CredentialProviderConfig, error) {
+	obj, gvk, err := codecs.UniversalDecoder().Decode(data, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	if gvk.Kind != "CredentialProviderConfig" {
+		return nil, fmt.Errorf("failed to decode %q (wrong Kind)", gvk.Kind)
+	}
+
+	if gvk.Group != kubeletconfig.GroupName {
+		return nil, fmt.Errorf("failed to decode CredentialProviderConfig, unexpected Group: %s", gvk.Group)
+	}
+
+	if internalConfig, ok := obj.(*kubeletconfig.CredentialProviderConfig); ok {
+		return internalConfig, nil
+	}
+
+	return nil, fmt.Errorf("unable to convert %T to *CredentialProviderConfig", obj)
+}
+
+// validateCredentialProviderConfig validates CredentialProviderConfig.
+func validateCredentialProviderConfig(config *kubeletconfig.CredentialProviderConfig) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if len(config.Providers) == 0 {
+		allErrs = append(allErrs, field.Required(field.NewPath("providers"), "at least 1 item in plugins is required"))
+	}
+
+	fieldPath := field.NewPath("providers")
+	for _, provider := range config.Providers {
+		if strings.Contains(provider.Name, "/") {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("name"), provider.Name, "provider name cannot contain '/'"))
+		}
+
+		if strings.Contains(provider.Name, " ") {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("name"), provider.Name, "provider name cannot contain spaces"))
+		}
+
+		if provider.Name == "." {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("name"), provider.Name, "provider name cannot be '.'"))
+		}
+
+		if provider.Name == ".." {
+			allErrs = append(allErrs, field.Invalid(fieldPath.Child("name"), provider.Name, "provider name cannot be '..'"))
+		}
+
+		if provider.APIVersion == "" {
+			allErrs = append(allErrs, field.Required(fieldPath.Child("apiVersion"), "apiVersion is required"))
+		}
+
+		if _, ok := apiVersions[provider.APIVersion]; !ok {
+			validAPIVersions := []string{}
+			for apiVersion := range apiVersions {
+				validAPIVersions = append(validAPIVersions, apiVersion)
+			}
+
+			allErrs = append(allErrs, field.NotSupported(fieldPath.Child("apiVersion"), provider.APIVersion, validAPIVersions))
+		}
+
+		if len(provider.MatchImages) == 0 {
+			allErrs = append(allErrs, field.Required(fieldPath.Child("matchImages"), "at least 1 item in matchImages is required"))
+		}
+
```

Comment:
pass all the matchers through the same `ParseSchemelessURL` that `URLsMatchStr` will, so a malformed match string fails startup rather than failing at runtime

Reply:
done